common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,930 @@
+
+  static constexpr bool FastByteClear = Size == 8;
+
+  using MatchIndex =
+#if CARBON_X86_SIMD_SUPPORT
+      BitIndex<uint32_t, /*ByteEncoding=*/false, /*ZeroMask=*/0xFFFF0000>;
+#else
+      BitIndex<uint64_t, /*ByteEncoding=*/true>;
+#endif
+
+  using MatchRange = BitIndexRange<MatchIndex>;
+
+  union {
+    uint8_t bytes[Size];
+    uint64_t byte_ints[Size / 8];
+#if CARBON_NEON_SIMD_SUPPORT
+    uint8x8_t byte_vec = {};
+    static_assert(sizeof(byte_vec) == Size);
+#elif CARBON_X86_SIMD_SUPPORT
+    __m128i byte_vec = {};
+    static_assert(sizeof(byte_vec) == Size);
+#endif
+  };
+
+  auto Print(llvm::raw_ostream& out) const -> void;
+
+  friend auto operator==(MetadataGroup lhs, MetadataGroup rhs) -> bool {
+    return CompareEqual(lhs, rhs);
+  }
+
+  // The main API for this class. This API will switch between a portable and
+  // SIMD implementation based on what is most efficient, but in debug builds
+  // will cross check that the implementations do not diverge.
+  static auto Load(const uint8_t* metadata, ssize_t index) -> MetadataGroup;
+  auto Store(uint8_t* metadata, ssize_t index) const -> void;
+
+  auto ClearByte(ssize_t byte_index) -> void;
+
+  auto ClearDeleted() -> void;
+
+  auto Match(uint8_t present_byte) const -> MatchRange;
+  auto MatchPresent() const -> MatchRange;
+
+  auto MatchEmpty() const -> MatchIndex;
+  auto MatchDeleted() const -> MatchIndex;
+
+ private:
+  // Two classes only defined in the benchmark code are allowed to directly call
+  // the portable and SIMD implementations for benchmarking purposes.
+  friend class BenchmarkPortableMetadataGroup;
+  friend class BenchmarkSIMDMetadataGroup;
+
+  // Whether to use a SIMD implementation. Even when we *support* a SIMD
+  // implementation, we do not always have to use it in the event that it is
+  // less efficient than the portable version.
+  static constexpr bool UseSIMD =
+#if CARBON_X86_SIMD_SUPPORT
+      true;
+#else
+      false;
+#endif
+
+  // All SIMD variants that we have an implementation for should be enabled for
+  // debugging. This lets us maintain a SIMD implementation even if it is not
+  // used due to performance reasons, and easily re-enable it if the performance
+  // changes.
+  static constexpr bool DebugSIMD =
+#if !defined(NDEBUG) && (CARBON_NEON_SIMD_SUPPORT || CARBON_X86_SIMD_SUPPORT)
+      true;
+#else
+      false;
+#endif
+
+  static constexpr uint64_t MSBs = 0x8080'8080'8080'8080ULL;
+  static constexpr uint64_t LSBs = 0x0101'0101'0101'0101ULL;
+
+  using MatchMask = MatchIndex::MaskT;
+
+  static auto CompareEqual(MetadataGroup lhs, MetadataGroup rhs) -> bool;
+
+  // Directly verify the provided masks that will be used to build a
+  // `MatchIndex` or `MatchGroup`. These either `CHECK`-fail or return true.
+  auto VerifyIndexMask(
+      MatchMask index_mask,
+      llvm::function_ref<auto(uint8_t mask_byte)->bool> byte_match) const
+      -> bool;
+  auto VerifyRangeMask(
```

Comment:
I took a stab. Note that the comment use to refer to `MatchGroup` which I think is now `MatchRange`.
  
```suggestion
  // Functions for validating the returned matches agree with what is predicted by the `byte_match` function. These either `CHECK`-fail or return true.
  // To pass validation, the `*_mask` argument must have `0x80` for those bytes where `byte_match` returns true, and `0` for the rest.

  // `VerifyIndexMask` is for functions that return `MatchIndex`, as they only promise to return accurate information up to the first match.
  auto VerifyIndexMask(
      MatchMask index_mask,
      llvm::function_ref<auto(uint8_t mask_byte)->bool> byte_match) const
      -> bool;
  // `VerifyRangeMask` is for functions that return `MatchRange`, and so it validates all the bytes of `range_mask`.
  auto VerifyRangeMask(
```